cmd/fs-rwpool.go

```diff
@@ -0,0 +1,101 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"os"
+	pathutil "path"
+	"sync"
+
+	"github.com/minio/minio/pkg/lock"
+)
+
+// fsIOPool represents a protected list to keep track of all
+// the concurrent readers at a given path.
+type fsIOPool struct {
+	sync.Mutex
+	readersMap map[string]*lock.RLockedFile
+}
+
+// Open is a wrapper call to read locked file which
+// returns a ReadAtCloser.
+//
+// ReaderAt is provided so that the fd is non seekable, since
+// we are sharing fd's with concurrent threads, we don't want
+// all readers to change offsets on each other during such
+// concurrent operations. Using ReadAt allows us to read from
+// any offsets.
+//
+// Closer is implemented to track total readers and to close
+// only when there no more readers, the fd is purged if the lock
+// count has reached zero.
+func (fsi *fsIOPool) Open(path string) (*lock.RLockedFile, error) {
+	fsi.Lock()
+	defer fsi.Unlock()
+	rlkFile, ok := fsi.readersMap[path]
+	// File reference exists on map, validate if its
+	// really closed and we are safe to purge it.
+	if ok {
+		// If the file is closed, remove it from the reader pool map.
+		if rlkFile.IsClosed() {
+			// Purge the cached lock path from map.
+			delete(fsi.readersMap, path)
+			// Indicate that we can populate the new fd.
+			ok = false
+		} else {
+			// Increment the lock ref, since the file is not closed yet
+			// and caller requested to read the file again.
+			rlkFile.IncLockRef()
+		}
+	}
+	// Locked path reference doesn't exist, freshly open the file in
+	// read lock mode.
+	if !ok {
+		var err error
+		rlkFile, err = lock.RLockedOpenFile(preparePath(path))
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, errFileNotFound
+			} else if os.IsPermission(err) {
+				return nil, errFileAccessDenied
+			}
+			return nil, err
+		}
+		fsi.readersMap[path] = rlkFile
+	}
+	return rlkFile, nil
+}
+
+// Create - creates a new write locked file instance.
+// Attempt to lock the file if it exists,
+// - if the file doesn't exist. We create the file and hold lock.
+// - if the file exists. Then we try to get a write lock this
+//   will block if we can't get a lock perhaps another write
+//   or read is in progress. Concurrent calls are protected
+//   by the global namspace lock within the same process.
+func (fsi *fsIOPool) Create(path string) (wlk *lock.RWLockedFile, err error) {
+	if err = mkdirAll(pathutil.Dir(path), 0777); err != nil {
+		return nil, err
+	}
+	wlk, err = lock.RWLockedOpenFile(preparePath(path))
+	if err != nil {
+		if err = os.Remove(preparePath(pathutil.Dir(path))); err != nil {
```

Comment:
it's safer not to remove in case of error so that already existing file is not removed because of a temporary IO error etc